File: test/conformance/c/rows/E5.c
/* test/conformance/c/rows/E5.c — schema matrix cell c/E5: "?T vs plain
 * nesting" (docs/roadmap.sexp:2313, matrix schema#876).
 *
 * THE LAW (docs/SPEC-TABLES.md:7194): "A field moved between `?T` and a plain
 * nesting is not an evolution event at all — the bytes do not move." P1 nests
 * Link BY VALUE (test/tables/P1.schema); P3 marks the same field `?Link`
 * (test/tables/P3.schema); on the tolerant table wire the two spellings are
 * byte-identical for content that is not entirely default
 * (test/tables/main.cpp, "the three spellings over NON-DEFAULT content"), so
 * every direction of the seam reads SILENT — no counter moved, no refusal,
 * the payload rides whole. The asymmetry at the empty end is pinned too: a
 * by-value nesting at its defaults writes nothing while a PRESENT optional
 * writes its body anyway, so the round-trip byte identity is asserted on the
 * non-default directions only.
 *
 * The conformance corpus pins the read report of every direction in
 * testdata/conformance/tables/reports.txt — p1_as_p3, p3_as_p1,
 * p1_empty_as_p3, p3_empty_as_p1 all answer 0,0,0,0,0,false,read — and the
 * wire bytes live in testdata/wire/tables/. This test walks the SAME path the
 * C driver does (test/conformance/c/main.c): the erased ConformanceCodec each
 * unit's translation unit exports, over the generated C tables code. It adds
 * the byte-level half the report surface does not look at — read a file
 * written under one schema's layout and save it back under the other's; the
 * bytes must be identical (docs/FIXED-FORM-ALGORITHM.md:1683, proof item 1).
 *
 * Standalone: depends on no other rows/ file, links only the two P units'
 * translation units and their generated tables, edits no shared file. */

#include "driver.h"

#include <stdio.h>

static int failed;

static void check( int ok, const char * what )
{
    printf( "%s: %s\n", ok ? "ok" : "FAIL", what );
    if ( !ok ) { failed++; }
}

static const ConformanceCodec * find_codec( const ConformanceCodec * codecs, int count,
                                            const char * unit, const char * root )
{
    int i;
    for ( i = 0; i < count; i++ )
    {
        if ( strcmp( codecs[i].unit, unit ) == 0 && strcmp( codecs[i].root, root ) == 0 )
        {
            return &codecs[i];
        }
    }
    return NULL;
}

static uint8_t * read_file( const char * path, size_t * bytes )
{
    FILE * file = fopen( path, "rb" );
    long size;
    uint8_t * out;
    if ( file == NULL ) { return NULL; }
    fseek( file, 0, SEEK_END );
    size = ftell( file );
    fseek( file, 0, SEEK_SET );
    out = (uint8_t *) malloc( (size_t) ( size > 0 ? size : 1 ) );
    if ( out == NULL ) { fclose( file ); return NULL; }
    if ( size > 0 && fread( out, 1, (size_t) size, file ) != (size_t) size )
    {
        free( out );
        fclose( file );
        return NULL;
    }
    fclose( file );
    *bytes = (size_t) size;
    return out;
}

static int silent( const ConformanceReport * r )
{
    return r->unknown == 0 && r->kind_mismatch == 0 && r->widened == 0
        && r->clamped == 0 && r->duplicate == 0 && !r->malformed && !r->refused;
}

/* Read <wire_path>, written under the OTHER spelling of the seam, with this
 * codec's schema and re-save under it. Silent is the law's "not an evolution
 * event"; round-trip byte identity is the law's "the bytes do not move". */
static void cross_read( const ConformanceCodec * codec, const char * wire_path,
                        const char * what, int require_identical )
{
    uint8_t * wire = NULL;
    size_t wire_bytes = 0;
    void * value;
    ConformanceReport report;
    uint8_t * out = NULL;
    int64_t size;
    char line[512];

    wire = read_file( wire_path, &wire_bytes );
    if ( wire == NULL )
    {
        snprintf( line, sizeof( line ), "%s: cannot read %s", what, wire_path );
        check( 0, line );
        return;
    }
    value = codec->storage();
    memset( &report, 0, sizeof( report ) );
    if ( !codec->load( value, wire, (int64_t) wire_bytes, &report ) )
    {
        snprintf( line, sizeof( line ), "%s: the read failed outright", what );
        check( 0, line );
        free( wire );
        return;
    }
    snprintf( line, sizeof( line ), "%s: silent cross-read (unknown %d kind_mismatch %d widened %d clamped %d duplicate %d malformed %d refused %d)",
              what, report.unknown, report.kind_mismatch, report.widened,
              report.clamped, report.duplicate, report.malformed, report.refused );
    check( silent( &report ), line );

    if ( require_identical )
    {
        int ok;
        size = codec->measure( value );
        if ( size < 0 ) { check( 0, "cross_read: measure failed" ); }
        else
        {
            out = (uint8_t *) malloc( (size_t) size );
            if ( out == NULL ) { check( 0, "cross_read: out of memory" ); }
            else
            {
                ok = codec->save( value, out, size ) == size
                    && (uint64_t) size == wire_bytes
                    && memcmp( out, wire, (size_t) size ) == 0;
                snprintf( line, sizeof( line ), "%s: re-save is byte-identical (%lld/%zu bytes)",
                          what, (long long) size, wire_bytes );
                check( ok, line );
                free( out );
            }
        }
    }
    free( wire );
}

int main( void )
{
    int count;
    const ConformanceCodec * p1 = NULL;
    const ConformanceCodec * p3 = NULL;

    p1 = find_codec( conformance_codecs_tblp1( &count ), count, "tblp1", "Chain" );
    p3 = find_codec( conformance_codecs_tblp3( &count ), count, "tblp3", "Chain" );
    check( p1 != NULL && p3 != NULL, "both P codecs (tblp1/Chain, tblp3/Chain) are on the driver's table" );

    if ( p1 != NULL && p3 != NULL )
    {
        /* P1's plain-nested file, read under P3's ?Link schema: silent, and
           the bytes come back unchanged. P3 declares the field optional and
           its save writes the identical file (test/tables/main.cpp pins the
           two spellings byte-identical). */
        cross_read( p3, "testdata/wire/tables/chain_value.bin", "p1_as_p3", 1 );

        /* P3's optional-present file, read under P1's plain schema: silent,
           and the bytes come back unchanged. */
        cross_read( p1, "testdata/wire/tables/chain_optional.bin", "p3_as_p1", 1 );

        /* P1's all-default file, read under P3: silent; the empty end elides
           in both spellings, so the round trip is still byte-identical. */
        cross_read( p3, "testdata/wire/tables/chain_value_empty.bin", "p1_empty_as_p3", 1 );

        /* P3's present-and-all-default file, read under P1: silent (the body
           rides), but re-saving under P1 elides — the documented asymmetry —
           so no byte identity is demanded here. */
        cross_read( p1, "testdata/wire/tables/chain_optional_empty.bin", "p3_empty_as_p1", 0 );
    }

    printf( "E5 %s\n", failed ? "RED" : "GREEN" );
    return failed ? 1 : 0;
}

File: test/conformance/java/rows/E5.java
// java/E5 — "?T vs plain nesting"
//
// docs/FIXED-FORM-ALGORITHM.md:182 — "?T: 1 + C(T) — the present flag THEN the
// payload, which rides WHOLE"
// docs/SPEC-TABLES.md:6556-6565 — "KIND 35 IS THE ONE KIND THIS LAYOUT ADDS
// TO §3'S CLOSED SET... its size is one present byte plus its child's... ?T
// and a plain T nesting are WIRE-IDENTICAL... ON THIS FORM THEY ARE ONE BYTE
// APART"
//
// THE FIXTURE: P1.schema nests Link by value (plain T), P3.schema marks the
// same field `?Link` (optional T). On form 3 the two layouts differ by the
// present flag, and the kind changes from 13 (struct) to 35 (optional wrapper).
// This test verifies both formats write and read correctly.
//
// Run: java -cp build/conformance-java test/conformance/java/rows/E5.java
// exit 0 green, exit 1 red, one printed line per assertion.

public final class E5 {
    private E5() {}

    private static int failures = 0;

    private static void check(String what, boolean ok) {
        System.out.println(ok ? "ok   " + what : "FAILED " + what);
        if (!ok) { failures++; }
    }

    // layout entry: struct Entry { id : u64 ; kind : u8 ; size : u32 ; children : u32 } (17 bytes)
    // docs/FIXED-FORM-ALGORITHM.md:33
    private static int u32(byte[] b, int at) {
        return (b[at] & 0xff) | ((b[at + 1] & 0xff) << 8)
                | ((b[at + 2] & 0xff) << 16) | ((b[at + 3] & 0xff) << 24);
    }

    private static int layoutKind(byte[] layout, int entryIndex) {
        return layout[4 + entryIndex * 17 + 8] & 0xff;
    }

    private static int layoutSize(byte[] layout, int entryIndex) {
        return u32(layout, 4 + entryIndex * 17 + 9);
    }

    public static void main(String[] args) {
        // --- P1: plain Link nesting ---
        check("P1 ChainFixed bodyBytes == 36", tblp1.ChainFixed.bodyBytes == 36);
        check("P1 ChainFixed link entry kind == 13 (struct)",
                layoutKind(tblp1.ChainFixed.layout, 2) == 13);
        check("P1 ChainFixed link entry size == 16",
                layoutSize(tblp1.ChainFixed.layout, 2) == 16);

        // --- P3: ?Link optional nesting ---
        check("P3 ChainFixed bodyBytes == 37", tblp3.ChainFixed.bodyBytes == 37);
        check("P3 ChainFixed optional wrapper kind == 35",
                layoutKind(tblp3.ChainFixed.layout, 2) == 35);
        check("P3 ChainFixed optional wrapper size == 17 (1 present + 16 payload)",
                layoutSize(tblp3.ChainFixed.layout, 2) == 17);
        check("P3 ChainFixed nested Link entry kind == 13",
                layoutKind(tblp3.ChainFixed.layout, 3) == 13);

        // --- P1 write-then-read round-trip ---
        tblp1.ChainFixed.Value p1v = new tblp1.ChainFixed.Value();
        p1v.nameLength = 4;
        p1v.name[0] = 'T'; p1v.name[1] = 'e'; p1v.name[2] = 's'; p1v.name[3] = 't';
        p1v.link.value = 42;
        p1v.link.tagLength = 3;
        p1v.link.tag[0] = 'a'; p1v.link.tag[1] = 'b'; p1v.link.tag[2] = 'c';
        int need = tblp1.ChainFixed.measure(1);
        byte[] p1buf = new byte[need];
        int wrote = tblp1.ChainFixed.save(new tblp1.ChainFixed.Value[] { p1v }, 1, p1buf);
        check("P1 save returns measure", wrote == need);
        tblp1.ChainFixed.Value p1back = new tblp1.ChainFixed.Value();
        tblp1.TableFixed.Report r1 = new tblp1.TableFixed.Report();
        int n1 = tblp1.ChainFixed.load(
            new tblp1.ChainFixed.Value[] { p1back }, 1, p1buf,
            tblp1.TableFixed.plan(8192), new short[8192], tblp1.ChainFixed.image(), r1);
        check("P1 load returns 1", n1 == 1);
        check("P1 load no refusal", !r1.refused);
        check("P1 load name", p1back.nameLength == 4
                && p1back.name[0] == 'T' && p1back.name[1] == 'e'
                && p1back.name[2] == 's' && p1back.name[3] == 't');
        check("P1 load link value", p1back.link.value == 42);
        check("P1 load link tag", p1back.link.tagLength == 3
                && p1back.link.tag[0] == 'a'
                && p1back.link.tag[1] == 'b'
                && p1back.link.tag[2] == 'c');

        // --- P3 write-then-read round-trip (present) ---
        tblp3.ChainFixed.Value p3v = new tblp3.ChainFixed.Value();
        p3v.nameLength = 4;
        p3v.name[0] = 'T'; p3v.name[1] = 'e'; p3v.name[2] = 's'; p3v.name[3] = 't';
        p3v.linkPresent = true;
        p3v.link.value = 42;
        p3v.link.tagLength = 3;
        p3v.link.tag[0] = 'a'; p3v.link.tag[1] = 'b'; p3v.link.tag[2] = 'c';
        need = tblp3.ChainFixed.measure(1);
        byte[] p3buf = new byte[need];
        wrote = tblp3.ChainFixed.save(new tblp3.ChainFixed.Value[] { p3v }, 1, p3buf);
        check("P3 present save returns measure", wrote == need);
        tblp3.ChainFixed.Value p3back = new tblp3.ChainFixed.Value();
        tblp3.TableFixed.Report r3 = new tblp3.TableFixed.Report();
        int n3 = tblp3.ChainFixed.load(
            new tblp3.ChainFixed.Value[] { p3back }, 1, p3buf,
            tblp3.TableFixed.plan(8192), new short[8192], tblp3.ChainFixed.image(), r3);
        check("P3 present load returns 1", n3 == 1);
        check("P3 present load no refusal", !r3.refused);
        check("P3 present load linkPresent", p3back.linkPresent);
        check("P3 present load link value", p3back.link.value == 42);

        // --- P3 write-then-read round-trip (absent) ---
        tblp3.ChainFixed.Value p3a = new tblp3.ChainFixed.Value();
        p3a.nameLength = 4;
        p3a.name[0] = 'T'; p3a.name[1] = 'e'; p3a.name[2] = 's'; p3a.name[3] = 't';
        p3a.linkPresent = false;
        need = tblp3.ChainFixed.measure(1);
        byte[] p3abuf = new byte[need];
        wrote = tblp3.ChainFixed.save(new tblp3.ChainFixed.Value[] { p3a }, 1, p3abuf);
        check("P3 absent save returns measure", wrote == need);
        tblp3.ChainFixed.Value p3aback = new tblp3.ChainFixed.Value();
        tblp3.TableFixed.Report r3a = new tblp3.TableFixed.Report();
        n3 = tblp3.ChainFixed.load(
            new tblp3.ChainFixed.Value[] { p3aback }, 1, p3abuf,
            tblp3.TableFixed.plan(8192), new short[8192], tblp3.ChainFixed.image(), r3a);
        check("P3 absent load returns 1", n3 == 1);
        check("P3 absent load no refusal", !r3a.refused);
        check("P3 absent load linkPresent false", !p3aback.linkPresent);
        check("P3 absent load link value == 0", p3aback.link.value == 0);
        check("P3 absent load link tagLength == 0", p3aback.link.tagLength == 0);

        // --- Byte-level check: P3 body is one byte larger than P1 ---
        check("P3 recordBytes == P1 recordBytes + 1",
                tblp3.ChainFixed.recordBytes == tblp1.ChainFixed.recordBytes + 1);

        // --- For a present optional, both formats carry the same name bytes ---
        // Both records have name at offsets 0-19 (4-byte length + 16 bytes text)
        int p1RecordAt = tblp1.ChainFixed.headerBytes + 8;
        int p3RecordAt = tblp3.ChainFixed.headerBytes + 8;
        boolean sameName = true;
        for (int i = 0; i < 20; i++) {
            if (p1buf[p1RecordAt + i] != p3buf[p3RecordAt + i]) {
                sameName = false;
                break;
            }
        }
        check("P1 and P3 have same name bytes", sameName);

        if (failures != 0) {
            System.out.println(failures + " failure(s)");
            System.exit(1);
        }
        System.out.println("E5 green: ?T vs plain nesting on the Java leg");
    }
}